# Tidy Data
__In this notebook__:

Practice __tidy data__ fundamentals:
1. Data is tabular (rows and columns)
1. Each row is an observation
1. Each column in an attribute
1. Each cell has a single value

Use __pandas__ methods to shape the datasets:
1. `melt()`
1. `.drop()`
1. `.name`
1. `pivot_table()`
1. `reset_index()`

In [6]:
import pandas as pd
import numpy as np

## 1. __Attendance Data__

Load the `attendance.csv` file

In [7]:
df_attendance = pd.read_csv('./untidy-data/attendance.csv')

In [8]:
df_attendance.head()

Unnamed: 0 2018-01-01 2018-01-02 2018-01-03 2018-01-04 2018-01-05  \
0      Sally          P          T          T          H          P   
1       Jane          A          P          T          T          T   
2      Billy          A          T          A          A          H   
3       John          P          T          H          P          P   

  2018-01-06 2018-01-07 2018-01-08  
0          A          T          T  
1          T          A          T  
2          T          P          T  
3          T          P          P

In [9]:
df_attendance.replace(to_replace={'P': 1.00,
                                  'T': 0.90,
                                  'H': 0.50,
                                  'A': 0.00},
                      inplace=True)
# df.column.apply(function)

In [10]:
df_attendance.head()

Unnamed: 0  2018-01-01  2018-01-02  2018-01-03  2018-01-04  2018-01-05  \
0      Sally         1.0         0.9         0.9         0.5         1.0   
1       Jane         0.0         1.0         0.9         0.9         0.9   
2      Billy         0.0         0.9         0.0         0.0         0.5   
3       John         1.0         0.9         0.5         1.0         1.0   

   2018-01-06  2018-01-07  2018-01-08  
0         0.0         0.9         0.9  
1         0.9         0.0         0.9  
2         0.9         1.0         0.9  
3         0.9         1.0         1.0

In [11]:
df_attendance = df_attendance.melt(id_vars='Unnamed: 0')

In [12]:
df_attendance.rename(
    columns =({"Unnamed: 0" : "student",
               "variable": "date",
               "value": "attendance"}),
    inplace=True
)

In [13]:
df_attendance.head()

student        date  attendance
0   Sally  2018-01-01         1.0
1    Jane  2018-01-01         0.0
2   Billy  2018-01-01         0.0
3    John  2018-01-01         1.0
4   Sally  2018-01-02         0.9

In [14]:
df_attendance.sort_values(by=['student', 'date'], inplace=True)

In [15]:
df_attendance.head()

student        date  attendance
2    Billy  2018-01-01         0.0
6    Billy  2018-01-02         0.9
10   Billy  2018-01-03         0.0
14   Billy  2018-01-04         0.0
18   Billy  2018-01-05         0.5

In [16]:
df_attendance = df_attendance.reset_index(drop=True)

Calculate an `attendance` percentage for each student:
1. One half day is worth 50% of a full day
1. 10 tardies is equal to one absence.

In [17]:
df_attendance.head()

student        date  attendance
0   Billy  2018-01-01         0.0
1   Billy  2018-01-02         0.9
2   Billy  2018-01-03         0.0
3   Billy  2018-01-04         0.0
4   Billy  2018-01-05         0.5

In [18]:
attendance_grades = df_attendance.groupby(by=['student']).mean()

In [19]:
attendance_grades

attendance
student            
Billy        0.5250
Jane         0.6875
John         0.9125
Sally        0.7625

In [20]:
df_attendance.groupby(by=['student']).describe().T

student              Billy      Jane      John     Sally
attendance count  8.000000  8.000000  8.000000  8.000000
           mean   0.525000  0.687500  0.912500  0.762500
           std    0.459036  0.425735  0.172689  0.346152
           min    0.000000  0.000000  0.500000  0.000000
           25%    0.000000  0.675000  0.900000  0.800000
           50%    0.700000  0.900000  1.000000  0.900000
           75%    0.900000  0.900000  1.000000  0.925000
           max    1.000000  1.000000  1.000000  1.000000

Takeways:
1. John was the only student without an Absence.
    - minimum value is 0.5 == 'Half Day'
2. John was the most consistent with his attendance.
    - lowest standard deviation: ~0.17
3. Billy's 50th percentile shows that his attendance points consistented of:
    - Alot of absences
    - Tardy
    - Half day
4. Interesting. Comparing the standard deviation alone, one is able to rank the students.
    - John, Sally, Jane, and Billy.
    - Maybe it's only for this dataset?

## 2. __Coffee Levels__

Read the `coffee_levels.csv` file.

In [21]:
df_coffee_levels = pd.read_csv('./untidy-data/coffee_levels.csv')

Transform the data so that each carafe is in it's own column.

In [22]:
df_coffee_levels.head()

hour coffee_carafe  coffee_amount
0     8             x       0.816164
1     9             x       0.451018
2    10             x       0.843279
3    11             x       0.335533
4    12             x       0.898291

In [23]:
df_coffee_levels = df_coffee_levels.pivot_table(
    values= 'coffee_amount',
    index = 'hour',
    columns= 'coffee_carafe',
)

In [24]:
df_coffee_levels.reset_index(inplace=True)

In [25]:
df_coffee_levels.rename = ''

In [26]:
df_coffee_levels

coffee_carafe  hour         x         y         z
0                 8  0.816164  0.189297  0.999264
1                 9  0.451018  0.521502  0.915990
2                10  0.843279  0.023163  0.144928
3                11  0.335533  0.235529  0.311495
4                12  0.898291  0.017009  0.771947
5                13  0.310711  0.997464  0.398520
6                14  0.507288  0.058361  0.864464
7                15  0.215043  0.144644  0.436364
8                16  0.183891  0.544676  0.280621
9                17  0.391560  0.594126  0.436677

In [27]:
df_coffee_levels.reset_index(inplace=True)

Is this the best shape for the data?

Yes. We are comparing three types of coffee and their usage at a given hour. But this is NOT a definition of tidy data.

I would add details to the data set:
1. What the quantity represents: Bags, weight of grounds (lbs?)
    - Answered by Zach

## 3. __Cake Recipes__

Read the `cake_recipes.csv` data. This data set contains cake tastiness scores for combinations of different recipes, oven rack positions, and oven temperatures. Tidy the data as necessary.

In [59]:
df_cake_recipes = pd.read_csv('./untidy-data/cake_recipes.csv')

In [54]:
df_cake_recipes.head()

recipe:position        225        250        275        300
0        a:bottom  61.738655  53.912627  74.414730  98.786784
1           a:top  51.709751  52.009735  68.576858  50.228470
2        b:bottom  57.095320  61.904369  61.196980  99.248541
3           b:top  82.455004  95.224151  98.594881  58.169349
4        c:bottom  96.470207  52.001358  92.893227  65.473084

In [60]:
df_cake_recipes = df_cake_recipes.melt(id_vars='recipe:position',
                                       var_name='temperature',
                                       value_name='tastiness_score'
                                      )

In [61]:
df_cake_recipes.head()

recipe:position temperature  tastiness_score
0        a:bottom         225        61.738655
1           a:top         225        51.709751
2        b:bottom         225        57.095320
3           b:top         225        82.455004
4        c:bottom         225        96.470207

In [62]:
df_recipe_rack_position = df_cake_recipes[
    'recipe:position'].str.split(":", expand=True)

In [63]:
recipe_rack_position.columns = ['recipe', 'rack_position']

In [65]:
df_cake_recipes = pd.concat(
    [df_cake_recipes,
     recipe_rack_position
    ], axis=1)

In [66]:
df_cake_recipes = df_cake_recipes.drop(columns='recipe:position')

In [67]:
df_cake_recipes = df_cake_recipes[['recipe',
                                   'rack_position',
                                   'temperature',
                                   'tastiness_score'
                                  ]]

In [68]:
df_cake_recipes.head()

recipe rack_position temperature  tastiness_score
0      a        bottom         225        61.738655
1      a           top         225        51.709751
2      b        bottom         225        57.095320
3      b           top         225        82.455004
4      c        bottom         225        96.470207

Which `recipe`, on average, is the best? recipe b

In [65]:
df_cake_recipes.groupby(by='recipe').mean().reset_index()

recipe  tastiness_score
0      a        63.922201
1      b        76.736074
2      c        75.874748
3      d        62.864844

Which oven `temperature`, on average, produces the best results? 275

In [85]:
temperature_mean_score = df_cake_recipes.groupby(
    by=['temperature']).mean().reset_index()

temperature_mean_score.iloc[ \
    temperature_mean_score.tastiness_score.argmax()]

temperature            275
tastiness_score    74.8868
Name: 2, dtype: object

Which combination of `recipe`, `rack position`, and `temperature` gives the best result? recipe b, bottom rack, 300 degrees

In [68]:
df_cake_recipes.iloc[df_cake_recipes.tastiness_score.argmax()]

recipe                   b
rack_position       bottom
temperature            300
tastiness_score    99.2485
Name: 26, dtype: object

Other methods to look at:
1. `stack`
2. `unstack`
3. `pivot`